### PR TITLE
Improve HLP module efficiency using multiple parse stages

### LINKED ISSUE #16741: Engine - Optimize High-Level Parser Performance by Splitting Parsing Process
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
|5.0.0 | Engine | Manager | Packages/Sources | OS version |

## Description
An opportunity has been identified to significantly improve the performance of our High-Level Parser (HLP). It is proposed to split the parsing process into two separate stages: one for semantic analysis and one for result construction.

Currently, in most cases, the 'Logpar' parser fails, but it is possible for some individual component parsers to succeed. Generating the result mapped to the event is a costly process in terms of time and resources. Therefore, it is proposed to split the process into two stages:

1. quick semantic analysis: perform a quick semantic analysis to determine if all individual parsers in 'Logpar' are successful.
2. Construction and validation of the result: if all individual parsers are successful, proceed to construct the result and validate its content.

This division will allow optimizing HLP performance by avoiding unnecessary result construction and validation when not all individual parsers are successful. In addition, this strategy can reduce processing time and improve overall system efficiency.
## Tasks
- [x] Study Parser combinators
- [x] Study Logpar 
- [x] Propose an Implementation
- [x] Benchmarks (HLP vs HLP3) and (HLP2 vs HLP3)
- [ ] Implement new parser, trace and result
- [ ] Implement new combinators
- [ ] Implement new builders of logpar
- [ ] Implement old parsers with new semantic
- [ ] Combinators Test
- [ ] Parser, trace and result test
- [ ] Parser test
- [ ] Apply RTR corrections
- [ ] Update Documentation
- [ ] QA validate

### PR DESCRIPTION
**Pull request by JavierBejMen**
|Related issue|
|---|
| Closes #16741 |

This PR introduces enhancements to optimize the processing times of HLP in failure use cases. Currently, the HLP parsers execute various type validations and heavy parsers, and the combinator relies on merging JSONs at each step, resulting in prolonged parse times when an expression fails.

This PR proposes the following improvements by splitting the parse into multiple stages:

- Syntactic parser: Introducing a lightweight parsing process.
- Semantic parser: Performing type validation and intensive tasks.
- Mapping: Updating the event with parse results.

To streamline the parsing process, each stage must succeed in all parsers for an expression to advance to the next stage. This ensures that tasks are only performed when necessary, eliminating unnecessary overhead.

Additionally, structures of the parser combinator have been optimized and simplified.

### CODE REVIEW

File: src/engine/source/hlp/src/parsers/alphanumeric.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    {
+        auto synR = synP(txt);
+        if (synR.failure())
+        {
+            return abs::makeFailure<ResultT>(synR.remaining(), name);
+        }
+        else
+        {
+            return abs::makeSuccess(SemToken {syntax::parsed(synR, txt), semP}, synR.remaining());
+        }
```
Comment (juliancnn) [inline src/engine/source/hlp/src/parsers/alphanumeric.cpp:60]:
```suggestion
        return abs::makeSuccess(SemToken {syntax::parsed(synR, txt), semP}, synR.remaining());
```

File: src/engine/source/hlp/src/parsers/between.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        auto synR = synP(txt);
+        if (synR.failure())
+        {
+            return abs::makeFailure<ResultT>(synR.remaining(), name);
+        }
+        else
+        {
+            return abs::makeSuccess(SemToken {syntax::parsed(synR, txt), semP}, synR.remaining());
+        }
```
Comment (juliancnn) [inline src/engine/source/hlp/src/parsers/between.cpp:87]: same

File: src/engine/source/hlp/src/parsers/encodings.cpp [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        if (i == 0)
+        {
+            return abs::makeFailure<syntax::ResultT>(input, {});
+        }
+
+        // Consume up to two padding characters
+        if (i < input.size())
+        {
+            if (input[i] == '=')
+            {
```
Comment (juliancnn) [inline src/engine/source/hlp/src/parsers/encodings.cpp:84]: :eye: 
```suggestion
        if (i < input.size() && if (input[i] == '=')
        {
```

File: src/engine/source/hlp/src/parsers/encodings.cpp [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    {
+        auto synR = synP(txt);
+        if (synR.failure())
+        {
+            return abs::makeFailure<ResultT>(synR.remaining(), name);
+        }
+        else
+        {
+            return abs::makeSuccess(SemToken {syntax::parsed(synR, txt), semP}, synR.remaining());
+        }
```
Comment (juliancnn) [inline src/engine/source/hlp/src/parsers/encodings.cpp:128]:
```suggestion
        return abs::makeSuccess(SemToken {syntax::parsed(synR, txt), semP}, synR.remaining());
```

File: src/engine/source/hlp/src/parsers/ignore.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        auto synR = synP(txt);
+        if (synR.failure())
+        {
+            return abs::makeFailure<ResultT>(synR.remaining(), name);
+        }
+        else
+        {
+            return abs::makeSuccess(SemToken {syntax::parsed(synR, txt), semP}, synR.remaining());
+        }
+    };
```
Comment (juliancnn) [inline src/engine/source/hlp/src/parsers/ignore.cpp:47]: same

File: src/engine/source/hlp/src/parsers/ip.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        }
+
+        if (targetField.empty())
+        {
+            return noMapper();
+        }
+        else
+        {
+            return getMapper(parsed, targetField);
+        }
```
Comment (juliancnn) [inline src/engine/source/hlp/src/parsers/ip.cpp:45]: same

**Review by juliancnn — CHANGES_REQUESTED** [review]

File: src/engine/benchmark/hlp/CMakeLists.txt [hunk 7]
```diff
@@ -1,6 +1,11 @@
-add_executable(hlp_benchmarks
-  hlp_bench.cpp
-)
+# add_executable(hlp_benchmarks
+#   hlp_bench.cpp
+# )
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/benchmark/hlp/CMakeLists.txt:3]: Removed

File: src/engine/source/cmds/src/test.cpp [hunk 8]
```diff
@@ -3,21 +3,21 @@
 #include <atomic>
 #include <chrono>
 #include <memory>
-#include <thread>
 #include <termios.h>
+#include <thread>
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/cmds/src/test.cpp:7]: Done

File: src/engine/source/hlp/include/hlp/abstractParser.hpp [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    Result(Result&& other) noexcept
+        : m_value(std::move(other.m_value))
+        , m_remaining(std::move(other.m_remaining))
+        , m_trace(std::move(other.m_trace))
+        , m_success(std::move(other.m_success))
+        , m_hasValue(std::move(other.m_hasValue))
+        , m_nested(std::move(other.m_nested))
+    {
+    }
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/hlp/include/hlp/abstractParser.hpp:82]: Done

File: src/engine/source/hlp/include/hlp/abstractParser.hpp [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+    Result(T&& value, std::string_view remaining, bool success, std::string_view trace, bool hasValue, Nested&& nested)
+        : m_value(std::move(value))
+        , m_remaining(remaining)
+        , m_success(success)
+        , m_trace(trace)
+        , m_hasValue(hasValue)
+        , m_nested(std::move(nested))
+    {
+    }
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/hlp/include/hlp/abstractParser.hpp:49]: Done

File: src/engine/source/hlp/include/hlp/abstractParser.hpp [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+auto makeSuccess(T&& value, std::string_view remaining)
+{
+    return Result<T>(std::move(value), remaining, true, {}, true);
+}
+
+template<typename T, typename... Nested>
+auto makeSuccess(T&& value, std::string_view remaining, Nested&&... nested)
+{
+    return Result<T>(std::move(value), remaining, true, {}, true, std::move(nested)...);
+}
```
[the remark this answers is not in the document]
Comment (JavierBejMen) [reply src/engine/source/hlp/include/hlp/abstractParser.hpp:113]: Using a single one with defaults will imply to write defaulted ones when calling the function depending of the use case

File: src/engine/source/hlp/src/parsers/alphanumeric.cpp (hunk 1 above)
Comment (JavierBejMen) [reply src/engine/source/hlp/src/parsers/alphanumeric.cpp:60]: Done

File: src/engine/source/hlp/src/parsers/between.cpp (hunk 2 above)
Comment (JavierBejMen) [reply src/engine/source/hlp/src/parsers/between.cpp:87]: Done

File: src/engine/source/hlp/src/parsers/encodings.cpp (hunk 3 above)
Comment (JavierBejMen) [reply src/engine/source/hlp/src/parsers/encodings.cpp:84]: Done

File: src/engine/source/hlp/src/parsers/encodings.cpp (hunk 4 above)
Comment (JavierBejMen) [reply src/engine/source/hlp/src/parsers/encodings.cpp:128]: Done

File: src/engine/source/hlp/src/parsers/ip.cpp (hunk 6 above)
Comment (JavierBejMen) [reply src/engine/source/hlp/src/parsers/ip.cpp:45]: Done

File: src/engine/source/hlp/src/syntax.hpp [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
+/**
+ * @brief Matches the given literal
+ *
+ * @param lit The literal to match
+ * @param caseSensitive If true, the literal is case sensitive
+ * @return Parser
+ */
+inline Parser literal(const std::string lit, bool caseSensitive = true)
+{
+    if (caseSensitive)
```
Comment (LucioDonda) [inline src/engine/source/hlp/src/syntax.hpp:209]: Is case insensitive use in some place? or are we able to use it ?

File: src/engine/source/hlp/src/syntax.hpp [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+}
+
+/**
+ * @brief Contains the combinators for the syntax parsers, this combinators do not extract any value and do not nest the
+ * results.
+ *
+ */
+namespace combinators
+{
+inline Parser operator&(const Parser& lhs, const Parser& rhs)
```
Comment (LucioDonda) [inline src/engine/source/hlp/src/syntax.hpp:38]: All this overload operators are still of use or should be replaced ? 
e.g. OR (|) by combinator::choice() ?

**Review by LucioDonda — APPROVED** [review]
LGTM

**Review by juliancnn — APPROVED** [review]
LGTM!

File: src/engine/source/hlp/src/syntax.hpp (hunk 13 above)
Comment (JavierBejMen) [reply src/engine/source/hlp/src/syntax.hpp:38]: it is used by some syntax parsers

**Review by JcabreraC — APPROVED** [review]
LGTM !

**Pull request merged**